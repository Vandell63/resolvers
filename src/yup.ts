import { Resolver, transformToNestObject } from 'react-hook-form';
import { FieldError, FieldErrors } from 'react-hook-form/dist/types/form';
import { DeepMap } from 'react-hook-form/dist/types/utils';
import Yup from 'yup';

const parseErrorSchema = <TFieldValues extends Record<string, any>>(
  error: Yup.ValidationError,
  validateAllFieldCriteria: boolean,
): DeepMap<TFieldValues, FieldError> =>
  Array.isArray(error.inner) && error.inner.length
    ? error.inner.reduce<FieldErrors<TFieldValues>>(
        (previous, { path, message, type }) => {
          const previousPath = previous[path] as FieldErrors | undefined;
          const previousTypes = validateAllFieldCriteria
            ? previousPath?.types || []
            : {};
          return {
            ...previous,
            [path]: {
              ...(previous[path] || {
                message,
                type,
                types: previousTypes,
              }),
              types: {
                ...previousTypes,
                [type]: validateAllFieldCriteria
                  ? [message, ...(previousTypes[type] || [])]
                  : message,
              },
            },
          };
        },
        {},
      )
    : ({
        [error.path]: {
          message: error.message,
          type: error.type,
        },
      } as DeepMap<TFieldValues, FieldError>);

export const yupResolver = <TFieldValues extends Record<string, any>>(
  schema: Yup.ObjectSchema | Yup.Lazy,
  options: Yup.ValidateOptions = {
    abortEarly: false,
  },
): Resolver<TFieldValues> => async (
  values,
  context,
  validateAllFieldCriteria = false,
) => {
  try {
    return {
      values: (await schema.validate(values, {
        context,
        ...options,
      })) as any,
      errors: {},
    };
  } catch (e) {
    return {
      values: {},
      errors: transformToNestObject(
        parseErrorSchema(e, validateAllFieldCriteria),
      ),
    };
  }
};
