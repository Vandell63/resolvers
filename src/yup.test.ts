import * as yup from 'yup';
import { yupResolver } from './yup';

const errors = {
  name: 'ValidationError',
  value: { createdOn: '2019-03-27T04:05:51.503Z' },
  path: undefined,
  type: undefined,
  errors: ['name is a required field', 'age is a required field'],
  inner: [
    {
      name: 'ValidationError',
      value: undefined,
      path: 'name',
      type: 'required',
      errors: [],
      inner: [],
      message: 'name is a required field',
      params: [],
    },
    {
      name: 'ValidationError',
      value: undefined,
      path: 'name',
      type: 'min',
      errors: [],
      inner: [],
      message: 'name is a min field',
      params: [],
    },
    {
      name: 'ValidationError',
      value: undefined,
      path: 'age',
      type: 'required',
      errors: [],
      inner: [],
      message: 'age is a required field',
      params: [],
    },
  ],
};

const schema = yup.object().shape({
  name: yup.string().required(),
  age: yup.number().required().positive().integer(),
  email: yup.string().email(),
  website: yup.string().url(),
  createdOn: yup.date().default(function () {
    return new Date();
  }),
  foo: yup
    .array()
    .required()
    .of(
      yup.object({
        loose: yup.boolean(),
      }),
    ),
});

describe('yupResolver', () => {
  it('should get values', async () => {
    const data = {
      name: 'jimmy',
      age: '24',
      createdOn: '2014-09-23T19:25:25Z',
      foo: [{ yup: true }],
    };
    expect(await yupResolver(schema)(data)).toEqual({
      errors: {},
      values: {
        name: 'jimmy',
        age: 24,
        foo: [{ yup: true }],
        createdOn: new Date('2014-09-23T19:25:25Z'),
      },
    });
  });

  it('should get errors', async () => {
    const data = {
      name: 2,
      age: 'test',
      createdOn: null,
      foo: [{ loose: null }],
    };
    expect(await yupResolver(schema)(data)).toMatchSnapshot();
  });

  it('should pass down the yup context', async () => {
    const data = { name: 'eric' };
    const context = { min: true };
    const schemaWithContext = yup.object().shape({
      name: yup
        .string()
        .required()
        .when('$min', (min: boolean, schema: yup.StringSchema) => {
          return min ? schema.min(6) : schema;
        }),
    });
    schemaWithContext.validate = jest.fn().mockResolvedValue({});
    await yupResolver(schemaWithContext)(data, context);
    expect(schemaWithContext.validate).toHaveBeenCalled();
    expect(schemaWithContext.validate).toHaveBeenCalledWith(data, {
      abortEarly: false,
      context,
    });
  });
});

describe('validateWithSchema', () => {
  it('should return undefined when no error reported', async () => {
    expect(
      await yupResolver({
        validate: () => {
          throw errors;
        },
      } as any)({}),
    ).toMatchSnapshot();
  });

  it('should return empty object when validate pass', async () => {
    expect(
      await yupResolver({
        validate: () => new Promise((resolve) => resolve()),
      } as any)({}),
    ).toEqual({
      errors: {},
      values: undefined,
    });
  });

  it('should return an error based on the user context', async () => {
    const data = { name: 'eric' };
    const schemaWithContext = yup.object().shape({
      name: yup
        .string()
        .required()
        .when('$min', (min: boolean, schema: yup.StringSchema) => {
          return min ? schema.min(6) : schema;
        }),
    });
    expect(await yupResolver(schemaWithContext)(data, { min: true }))
      .toMatchInlineSnapshot(`
      Object {
        "errors": Object {
          "name": Object {
            "message": "name must be at least 6 characters",
            "type": "min",
          },
        },
        "values": Object {},
      }
    `);
  });
});
