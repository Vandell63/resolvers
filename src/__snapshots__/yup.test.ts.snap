// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validateWithSchema should return undefined when no error reported 1`] = `
Object {
  "errors": Object {
    "age": Object {
      "message": "age is a required field",
      "type": "required",
      "types": Object {
        "required": "age is a required field",
      },
    },
    "name": Object {
      "message": "name is a required field",
      "type": "required",
      "types": Object {
        "min": "name is a min field",
      },
    },
  },
  "values": Object {},
}
`;

exports[`yupResolver errors should get errors with validate all criteria fields 1`] = `
Object {
  "errors": Object {
    "age": Object {
      "message": "age must be a \`number\` type, but the final value was: \`NaN\` (cast from the value \`\\"test\\"\`).",
      "type": "typeError",
      "types": Object {
        "typeError": Array [
          "age must be a \`number\` type, but the final value was: \`NaN\` (cast from the value \`\\"test\\"\`).",
        ],
      },
    },
    "createdOn": Object {
      "message": "createdOn must be a \`date\` type, but the final value was: \`Invalid Date\`.",
      "type": "typeError",
      "types": Object {
        "typeError": Array [
          "createdOn must be a \`date\` type, but the final value was: \`Invalid Date\`.",
        ],
      },
    },
    "foo": Array [
      Object {
        "loose": Object {
          "message": "foo[0].loose must be a \`boolean\` type, but the final value was: \`null\`.
 If \\"null\\" is intended as an empty value be sure to mark the schema as \`.nullable()\`",
          "type": "typeError",
          "types": Object {
            "typeError": Array [
              "foo[0].loose must be a \`boolean\` type, but the final value was: \`null\`.
 If \\"null\\" is intended as an empty value be sure to mark the schema as \`.nullable()\`",
            ],
          },
        },
      },
    ],
    "password": Object {
      "message": "password is a required field",
      "type": "required",
      "types": Object {
        "matches": Array [
          "Special",
          "Number",
          "Uppercase",
          "Lowercase",
        ],
        "min": Array [
          "password must be at least 8 characters",
        ],
        "required": Array [
          "password is a required field",
        ],
      },
    },
  },
  "values": Object {},
}
`;

exports[`yupResolver errors should get errors without validate all criteria fields 1`] = `
Object {
  "errors": Object {
    "age": Object {
      "message": "age must be a \`number\` type, but the final value was: \`NaN\` (cast from the value \`\\"test\\"\`).",
      "type": "typeError",
      "types": Object {
        "typeError": "age must be a \`number\` type, but the final value was: \`NaN\` (cast from the value \`\\"test\\"\`).",
      },
    },
    "createdOn": Object {
      "message": "createdOn must be a \`date\` type, but the final value was: \`Invalid Date\`.",
      "type": "typeError",
      "types": Object {
        "typeError": "createdOn must be a \`date\` type, but the final value was: \`Invalid Date\`.",
      },
    },
    "foo": Array [
      Object {
        "loose": Object {
          "message": "foo[0].loose must be a \`boolean\` type, but the final value was: \`null\`.
 If \\"null\\" is intended as an empty value be sure to mark the schema as \`.nullable()\`",
          "type": "typeError",
          "types": Object {
            "typeError": "foo[0].loose must be a \`boolean\` type, but the final value was: \`null\`.
 If \\"null\\" is intended as an empty value be sure to mark the schema as \`.nullable()\`",
          },
        },
      },
    ],
    "password": Object {
      "message": "password is a required field",
      "type": "required",
      "types": Object {
        "required": "password is a required field",
      },
    },
  },
  "values": Object {},
}
`;
